### PR TITLE
Bugfix: Stats engine reconciliation for missed remove events.

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

This is a quick fix for cases where stats engine might miss
'container removed' event from the docker task engine
event stream.

Without this fix, the stats engine will spin on
'docker stats' api until the container is removed.

This is mentioned as a quick fix because this is not addressing the
root-cause of why an event is sometimes being missed by the
stats engine.

There's also the case of 'container added' event being missed
which could cause metrics to be misrepresented.

This is also proposed as a fix for the issue #515.

There's a minor refactor of the task engine code where locks have been
added to protect api.Container status'es

### Implementation details

* `engine.GetInstanceMetrics()` is used to trigger a state reconciliation in stats engine
* `container.collect()` is used to check if the container is terminal when disconnected from `docker stats`

### Testing

- [X] Builds (`make release`)
- [X] Unit tests (`make short-test`) pass
- [X] Integration tests (`make test`) pass
- [X] Functional tests (`make run-functional-tests`) pass
- Manual tests
 - [X] Built an agent, where 'stats engine' would deliberately ignore 'container removed' events and verified that it's still reconcile and not spin as it used to before this fix.

New tests cover the changes: Yes

### Description for the changelog

Fixes an issue where stats engine could miss 'remove' events and indefinitely query stats for stopped containers
### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes (AWS employee)

### CODE REVIEW

File: agent/engine/docker_container_engine.go [hunk 1]
```diff
@@ -726,7 +726,7 @@ func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.
 	go func() {
 		statsErr := client.Stats(options)
 		if statsErr != nil {
-			seelog.Warnf("Error retrieving stats for container %s: %v", id, statsErr)
+			seelog.Debugf("Error retrieving stats for container %s: %v", id, statsErr)
```
Comment (richardpen) [inline agent/engine/docker_container_engine.go]: Would it be reasonable to add a check here, if we keep receiving the error of "no such container" here, can we check the container status and if the container has already stopped, we just close the stream?

Comment (aaithal) [inline agent/engine/docker_container_engine.go]: I decided against doing that and leaving it up to the caller to handle the error. Because, eventually, the stream does get closed irrespective of the error printed here. Go-dockerclient ensures that. I think it makes more sense for the error to be surfaced via an error channel so that the caller can take the requisite action upon receiving the error in the channel (this is not part of this PR as it requires more thought and time). Does that make sense?

**Comment by richardpen** [comment]
👍

File: agent/engine/docker_container_engine.go (hunk 1 above)
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: Why did you move this down to debug level?

File: agent/stats/container.go [hunk 2]
```diff
@@ -61,21 +62,58 @@ func (container *StatsContainer) collect() {
...
+			// We were disconnected from the stats stream.
+			// Check if the container is terminal. If it is, stop collecting metrics.
+			// We might sometimes miss events from docker task  engine and this helps
+			// in reconciling the state.
+			terminal, err := container.terminal()
+			if err != nil {
+				seelog.Warnf("Error determining if the container %s is terminal, stopping stats collection: %v", dockerID, err)
+				container.StopStatsCollection()
+			} else if terminal {
+				// Error determining if the container is terminal. This means that the container
```
Comment (samuelkarp) [inline agent/stats/container.go]: I think this comment is in the wrong place.

File: agent/stats/engine.go [hunk 3]
```diff
@@ -395,6 +418,23 @@ func (engine *DockerStatsEngine) getContainerMetricsForTask(taskArn string) ([]*
 	return containerMetrics, nil
 }
 
+func (engine *DockerStatsEngine) doRemoveContainer(container *StatsContainer, taskArn string) {
+	container.StopStatsCollection()
+	dockerID := container.containerMetadata.DockerID
+	delete(engine.tasksToContainers[taskArn], dockerID)
+	seelog.Debugf("Deleted container from tasks, id: %s", dockerID)
+
+	if len(engine.tasksToContainers[taskArn]) == 0 {
```
Comment (samuelkarp) [inline agent/stats/engine.go]: Looks like you're only handling the case where there are no containers in the task.  Do you need to handle the case where there are containers?

File: agent/engine/docker_container_engine.go (hunk 1 above)
Comment (aaithal) [inline agent/engine/docker_container_engine.go]: Since the error handling is not done here anyway, having a WARN message here wouldn't help I think. Even when the container dies, you get a host of 'context canceled' or 'no such container' error messages here. I think it makes more sense for that logic to be with `container.processStatsStream`. Does that make sense?

File: agent/stats/container.go (hunk 2 above)
Comment (aaithal) [inline agent/stats/container.go]: Thanks, will move.

File: agent/stats/engine.go (hunk 3 above)
Comment (aaithal) [inline agent/stats/engine.go]: There's no need to remove the entry for the task if its composed of more than 1 container, this would be handled when that condition becomes true.

File: agent/stats/container.go (hunk 2 above)
Comment (aaithal) [inline agent/stats/container.go]: Fixed.

File: agent/engine/docker_container_engine.go (hunk 1 above)
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: We're losing errors from the `client.Stats` call if we log them at debug.  `container.processStatsStream` doesn't log anything at WARN (and doesn't really have much in the way of important errors) and `container.collect` just logs at DEBUG.

Comment (aaithal) [inline agent/engine/docker_container_engine.go]: Will make it `Infof` so that we don't loose track of it.

**Comment by samuelkarp** [comment]
:+1:

File: agent/stats/engine.go [hunk 4]
```diff
@@ -77,6 +77,19 @@ func (resolver *DockerContainerMetadataResolver) ResolveTask(dockerID string) (*
...
 }
 
+// ResolveContainer resolves the api container object, given container id.
+func (resolver *DockerContainerMetadataResolver) ResolveContainer(dockerID string) (*api.DockerContainer, error) {
+	if resolver.dockerTaskEngine == nil {
+		return nil, fmt.Errorf("Docker task engine uninitialized")
+	}
+	container, found := resolver.dockerTaskEngine.State().ContainerById(dockerID)
+	if !found {
+		return nil, fmt.Errorf("Could not map docker id to container")
```
Comment (kiranmeduri) [inline agent/stats/engine.go]: will it be helpful to include dockerID? (here and above)

Comment (aaithal) [inline agent/stats/engine.go]: Sure, will add.

Comment (aaithal) [inline agent/stats/engine.go]: @kiranmeduri added container id in the error message
